### ISSUE TITLE
AddKeyPath adds bip32_derivs to both inputs and outputs in PSBT

### ISSUE THREAD

**Issue by molnard** [MetacoSA/NBitcoin#927]
CC [was complaining](https://github.com/zkSNACKs/WalletWasabi/issues/4460) about `PSBT` in case of multisig address and didn't sign. 

### Repro
1. Start Wasabi and go tx builder
2. Send to an external Multisig address
3. CC will fail to sign, `PSBT` outputs contain an extra field. 

### PSBT changes during the build

Before `psbt.AddKeyPath`

```
  "inputs": [
    {
      "index": 0,
      "partial_signatures": {
        "censored"
      },
      "witness_utxo": {
        "value": "0.xxxxxxx",
        "scriptPubKey": "0 censored"
      },
      "bip32_derivs": []
    }
  ],
  "outputs": [
    {
      "bip32_derivs": []
    }
  ]
```
After psbt.AddKeyPath

```
  "inputs": [
    {
      "index": 0,
      "partial_signatures": {
        "censored"
      },
      "witness_utxo": {
        "value": "0.xxxxxxx",
        "scriptPubKey": "0 censored"
      },
      "bip32_derivs": [
        {
          "pubkey": "PUBKEY A",
          "master_fingerprint": "FINGERPRINT A",
          "path": "PATH A"
        }
      ]
    }
  ],
  "outputs": [
    {
      "bip32_derivs": [
        {
          "pubkey": "PUBKEY A",
          "master_fingerprint": "FINGERPRINT A",
          "path": "PATH A"
        }
      ]
    }
  ]
```

There should not be an extra output. 

Wasabi PR to hack it around https://github.com/zkSNACKs/WalletWasabi/pull/4601

**Comment by nopara73** [comment]
A 2of5 mainnet address one can reproduce the incorrect behavior with: `bc1qeef3jecqytj8j2xnjzduf5mu9c6jsqwd4hmvyv2zw8hzpf7a47nqrws5sn`.

**Comment by NicolasDorier** [comment]
I do not see an extra output here

**Comment by NicolasDorier** [comment]
I don't see why AddKeyPath should not add the KeyPath to the outputs?

**Comment by nopara73** [comment]
@NicolasDorier If build a PSBT where `bc1qeef3jecqytj8j2xnjzduf5mu9c6jsqwd4hmvyv2zw8hzpf7a47nqrws5sn` is an output, then call `AddKeyPath` for an input, the function will not only add the `bip32_derivs` field to the input, but also to the output.  

This makes no sense, because `bc1qeef3jecqytj8j2xnjzduf5mu9c6jsqwd4hmvyv2zw8hzpf7a47nqrws5sn` is not part of the wallet and that's why Coldcard doesn't sign.  

More specifically I think there's a bug somewhere in this condition:

```cs
if ((scriptPubKey != null && coin.ScriptPubKey == scriptPubKey) ||
					((o.GetSignableCoin() ?? coin.TryToScriptCoin(pubkey)) is Coin c && txBuilder.IsCompatibleKeyFromScriptCode(pubkey, c.GetScriptCode())) ||
					  txBuilder.IsCompatibleKeyFromScriptCode(pubkey, coin.ScriptPubKey))
```

**Comment by NicolasDorier** [comment]
I don't understand your bug. I tried to repro:

This test pass as expected.
```csharp
[Fact]
[Trait("UnitTest", "UnitTest")]
public void Temp()
{
	var keys = Enumerable.Range(0, 5).Select(_ => new ExtKey()).ToArray();

	var accountKey = new KeyPath("84'/0'/0'");
	var accountKeys = keys.Select(k => k.Derive(accountKey).Neuter()).ToArray();
	var depositPath = new KeyPath("0/0");
	var changePath = new KeyPath("1/0");
	var depositRedeem = PayToMultiSigTemplate.Instance.GenerateScriptPubKey(2, accountKeys.Select(k => k.Derive(depositPath).PubKey).ToArray());
	var changeRedeem = PayToMultiSigTemplate.Instance.GenerateScriptPubKey(2, accountKeys.Select(k => k.Derive(changePath).PubKey).ToArray());

	var c1 = CreateCoin(depositRedeem);

	var builder = Network.RegTest.CreateTransactionBuilder();
	builder.ShuffleOutputs = false;
	builder.AddCoins(c1);
	builder.SendEstimatedFees(new FeeRate(1.0m));
	builder.Send(new Key().PubKey.GetScriptPubKey(ScriptPubKeyType.Segwit), Money.Coins(0.2m));
	builder.SetChange(changeRedeem.WitHash.ScriptPubKey);
	var psbt = builder.BuildPSBT(false);
	foreach (var k in accountKeys)
	{
		psbt.AddKeyPath(k, depositPath);
		psbt.AddKeyPath(k, changePath);
	}
	Assert.Empty(psbt.Outputs[0].HDKeyPaths);
	Assert.NotEmpty(psbt.Outputs[1].HDKeyPaths);
}

int i = 0;
private ScriptCoin CreateCoin(Script redeem)
{
	var outpoint = new OutPoint(uint256.Zero, i++);
	var txout = new TxOut(Money.Coins(1.0m), redeem.WitHash.ScriptPubKey);
	return new ScriptCoin(outpoint, txout, redeem);
}
```

**Comment by nopara73** [comment]
`builder.Send(new Key().PubKey.GetScriptPubKey(ScriptPubKeyType.Segwit), Money.Coins(0.2m));` -> `builder.Send(BitcoinAddress.Create("bc1qeef3jecqytj8j2xnjzduf5mu9c6jsqwd4hmvyv2zw8hzpf7a47nqrws5sn", Network.Main), Money.Coins(0.2m));` https://github.com/MetacoSA/NBitcoin/pull/928

**Comment by NicolasDorier** [comment]
Will make a new version in 30min, please remove your hack then, I think it is not correct in some edge cases.

**Comment by NicolasDorier** [comment]
Pushed 5.0.60 should be online on nugget soon.

**Comment by molnard** [comment]
Thx! Was this a trivial fix? Do I need to do some testings?

**Comment by NicolasDorier** [comment]
it was trivial fix yes as you can see in 6fb6a6e
I added your specific case in the test suite.
